### ISSUE TITLE
Better handling of Redis connection errors

### ISSUE THREAD

**Issue by martinmo** [coleifer/huey#165]
I'm using Huey as part of a Django application, which works great 99% of the time. During some robustness tests I've found that when Huey loses its connection to Redis, it doesn't recover from that automatically, even if the Redis instance comes back online. 

Ideally, huey should retry connecting *or* completely shutdown on a `ConnectionError` (which seems to be easier to implement). A process manager like supervisord or systemd can take care of restarting it (of course with rate limiting in case of a permanent connection outage).

Additionally, there is an annoying cosmetic error in case of a connection failure: a chain of (at least) 6 exceptions is dumped to the console for each worker :(

**Comment by coleifer** [comment]
Testing locally, the consumer seemed to pop right back online after turning off Redis and turning it back on. This is using the "simple" example code, which is included in `examples/simple`, then running `./cons.sh`:

```python
[2016-07-12 22:14:08,153] DEBUG:huey.consumer.Worker:Worker-1:No messages, sleeping for: 0.305902286254
[2016-07-12 22:14:08,460] DEBUG:huey.consumer.Worker:Worker-1:No messages, sleeping for: 0.351787629192
[2016-07-12 22:14:08,460] DEBUG:huey.consumer.Worker:Worker-2:No messages, sleeping for: 0.351787629192
[2016-07-12 22:14:08,812] DEBUG:huey.consumer.Worker:Worker-1:No messages, sleeping for: 0.404555773571
[2016-07-12 22:14:08,813] DEBUG:huey.consumer.Worker:Worker-2:No messages, sleeping for: 0.404555773571
[2016-07-12 22:14:09,218] DEBUG:huey.consumer.Worker:Worker-1:No messages, sleeping for: 0.465239139606
[2016-07-12 22:14:09,218] DEBUG:huey.consumer.Worker:Worker-2:No messages, sleeping for: 0.465239139606
[2016-07-12 22:14:09,684] DEBUG:huey.consumer.Worker:Worker-1:No messages, sleeping for: 0.535025010547
[2016-07-12 22:14:09,684] DEBUG:huey.consumer.Worker:Worker-2:No messages, sleeping for: 0.535025010547
[2016-07-12 22:14:10,220] DEBUG:huey.consumer.Worker:Worker-2:No messages, sleeping for: 0.615278762129
[2016-07-12 22:14:10,220] DEBUG:huey.consumer.Worker:Worker-1:No messages, sleeping for: 0.615278762129
[2016-07-12 22:14:10,837] DEBUG:huey.consumer.Worker:Worker-2:No messages, sleeping for: 0.707570576449
[2016-07-12 22:14:10,837] DEBUG:huey.consumer.Worker:Worker-1:No messages, sleeping for: 0.707570576449
[2016-07-12 22:14:11,551] ERROR:huey.consumer.Worker:Worker-1:Error reading from queue
Traceback (most recent call last):
  File "/home/charles/pypath/huey/consumer.py", line 94, in loop
    task = self.huey.dequeue()
  File "/home/charles/pypath/huey/api.py", line 225, in dequeue
    message = self._dequeue()
  File "/home/charles/pypath/huey/api.py", line 164, in inner
    wrap_exception(exc_class)
  File "/home/charles/pypath/huey/utils.py", line 18, in wrap_exception
    raise new_exc_class('%s: %s' % (exc_class.__name__, exc))
QueueReadException: ConnectionError: Error 111 connecting to localhost:6379. Connection refused.
[2016-07-12 22:14:11,552] ERROR:huey.consumer.Worker:Worker-2:Error reading from queue
Traceback (most recent call last):
  File "/home/charles/pypath/huey/consumer.py", line 94, in loop
    task = self.huey.dequeue()
  File "/home/charles/pypath/huey/api.py", line 225, in dequeue
    message = self._dequeue()
  File "/home/charles/pypath/huey/api.py", line 164, in inner
    wrap_exception(exc_class)
  File "/home/charles/pypath/huey/utils.py", line 18, in wrap_exception
    raise new_exc_class('%s: %s' % (exc_class.__name__, exc))
QueueReadException: ConnectionError: Error 111 connecting to localhost:6379. Connection refused.
[2016-07-12 22:14:11,552] DEBUG:huey.consumer.Worker:Worker-1:No messages, sleeping for: 0.813706162916
[2016-07-12 22:14:11,552] DEBUG:huey.consumer.Worker:Worker-2:No messages, sleeping for: 0.813706162916
[2016-07-12 22:14:12,372] ERROR:huey.consumer.Worker:Worker-2:Error reading from queue
Traceback (most recent call last):
  File "/home/charles/pypath/huey/consumer.py", line 94, in loop
    task = self.huey.dequeue()
  File "/home/charles/pypath/huey/api.py", line 225, in dequeue
    message = self._dequeue()
  File "/home/charles/pypath/huey/api.py", line 164, in inner
    wrap_exception(exc_class)
  File "/home/charles/pypath/huey/utils.py", line 18, in wrap_exception
    raise new_exc_class('%s: %s' % (exc_class.__name__, exc))
QueueReadException: ConnectionError: Error 111 connecting to localhost:6379. Connection refused.
[2016-07-12 22:14:12,372] DEBUG:huey.consumer.Worker:Worker-2:No messages, sleeping for: 0.935762087354
[2016-07-12 22:14:12,373] ERROR:huey.consumer.Worker:Worker-1:Error reading from queue
Traceback (most recent call last):
  File "/home/charles/pypath/huey/consumer.py", line 94, in loop
    task = self.huey.dequeue()
  File "/home/charles/pypath/huey/api.py", line 225, in dequeue
    message = self._dequeue()
  File "/home/charles/pypath/huey/api.py", line 164, in inner
    wrap_exception(exc_class)
  File "/home/charles/pypath/huey/utils.py", line 18, in wrap_exception
    raise new_exc_class('%s: %s' % (exc_class.__name__, exc))
QueueReadException: ConnectionError: Error 111 connecting to localhost:6379. Connection refused.
[2016-07-12 22:14:12,373] DEBUG:huey.consumer.Worker:Worker-1:No messages, sleeping for: 0.935762087354
[2016-07-12 22:14:13,314] ERROR:huey.consumer.Worker:Worker-2:Error reading from queue
Traceback (most recent call last):
  File "/home/charles/pypath/huey/consumer.py", line 94, in loop
    task = self.huey.dequeue()
  File "/home/charles/pypath/huey/api.py", line 225, in dequeue
    message = self._dequeue()
  File "/home/charles/pypath/huey/api.py", line 164, in inner
    wrap_exception(exc_class)
  File "/home/charles/pypath/huey/utils.py", line 18, in wrap_exception
    raise new_exc_class('%s: %s' % (exc_class.__name__, exc))
QueueReadException: ConnectionError: Error 111 connecting to localhost:6379. Connection refused.
[2016-07-12 22:14:13,315] ERROR:huey.consumer.Worker:Worker-1:Error reading from queue
Traceback (most recent call last):
  File "/home/charles/pypath/huey/consumer.py", line 94, in loop
    task = self.huey.dequeue()
  File "/home/charles/pypath/huey/api.py", line 225, in dequeue
    message = self._dequeue()
  File "/home/charles/pypath/huey/api.py", line 164, in inner
    wrap_exception(exc_class)
  File "/home/charles/pypath/huey/utils.py", line 18, in wrap_exception
    raise new_exc_class('%s: %s' % (exc_class.__name__, exc))
QueueReadException: ConnectionError: Error 111 connecting to localhost:6379. Connection refused.
[2016-07-12 22:14:13,315] DEBUG:huey.consumer.Worker:Worker-2:No messages, sleeping for: 1.07612640046
[2016-07-12 22:14:13,315] DEBUG:huey.consumer.Worker:Worker-1:No messages, sleeping for: 1.07612640046
[2016-07-12 22:14:14,396] ERROR:huey.consumer.Worker:Worker-1:Error reading from queue
Traceback (most recent call last):
  File "/home/charles/pypath/huey/consumer.py", line 94, in loop
    task = self.huey.dequeue()
  File "/home/charles/pypath/huey/api.py", line 225, in dequeue
    message = self._dequeue()
  File "/home/charles/pypath/huey/api.py", line 164, in inner
    wrap_exception(exc_class)
  File "/home/charles/pypath/huey/utils.py", line 18, in wrap_exception
    raise new_exc_class('%s: %s' % (exc_class.__name__, exc))
QueueReadException: ConnectionError: Error 111 connecting to localhost:6379. Connection refused.
[2016-07-12 22:14:14,397] DEBUG:huey.consumer.Worker:Worker-1:No messages, sleeping for: 1.23754536053
[2016-07-12 22:14:14,398] ERROR:huey.consumer.Worker:Worker-2:Error reading from queue
Traceback (most recent call last):
  File "/home/charles/pypath/huey/consumer.py", line 94, in loop
    task = self.huey.dequeue()
  File "/home/charles/pypath/huey/api.py", line 225, in dequeue
    message = self._dequeue()
  File "/home/charles/pypath/huey/api.py", line 164, in inner
    wrap_exception(exc_class)
  File "/home/charles/pypath/huey/utils.py", line 18, in wrap_exception
    raise new_exc_class('%s: %s' % (exc_class.__name__, exc))
QueueReadException: ConnectionError: Error 111 connecting to localhost:6379. Connection refused.
[2016-07-12 22:14:14,398] DEBUG:huey.consumer.Worker:Worker-2:No messages, sleeping for: 1.23754536053
[2016-07-12 22:14:15,640] ERROR:huey.consumer.Worker:Worker-1:Error reading from queue
Traceback (most recent call last):
  File "/home/charles/pypath/huey/consumer.py", line 94, in loop
    task = self.huey.dequeue()
  File "/home/charles/pypath/huey/api.py", line 225, in dequeue
    message = self._dequeue()
  File "/home/charles/pypath/huey/api.py", line 164, in inner
    wrap_exception(exc_class)
  File "/home/charles/pypath/huey/utils.py", line 18, in wrap_exception
    raise new_exc_class('%s: %s' % (exc_class.__name__, exc))
QueueReadException: ConnectionError: Error 111 connecting to localhost:6379. Connection refused.
[2016-07-12 22:14:15,641] DEBUG:huey.consumer.Worker:Worker-1:No messages, sleeping for: 1.4231771646
[2016-07-12 22:14:15,642] ERROR:huey.consumer.Worker:Worker-2:Error reading from queue
Traceback (most recent call last):
  File "/home/charles/pypath/huey/consumer.py", line 94, in loop
    task = self.huey.dequeue()
  File "/home/charles/pypath/huey/api.py", line 225, in dequeue
    message = self._dequeue()
  File "/home/charles/pypath/huey/api.py", line 164, in inner
    wrap_exception(exc_class)
  File "/home/charles/pypath/huey/utils.py", line 18, in wrap_exception
    raise new_exc_class('%s: %s' % (exc_class.__name__, exc))
QueueReadException: ConnectionError: Error 111 connecting to localhost:6379. Connection refused.
[2016-07-12 22:14:15,642] DEBUG:huey.consumer.Worker:Worker-2:No messages, sleeping for: 1.4231771646
Exception in thread Scheduler:
Traceback (most recent call last):
  File "/usr/lib/python2.7/threading.py", line 801, in __bootstrap_inner
    self.run()
  File "/usr/lib/python2.7/threading.py", line 754, in run
    self.__target(*self.__args, **self.__kwargs)
  File "/home/charles/pypath/huey/consumer.py", line 360, in _run
    consumer_process.loop()
  File "/home/charles/pypath/huey/consumer.py", line 227, in loop
    for task in self.huey.read_schedule(now):
  File "/home/charles/pypath/huey/api.py", line 323, in read_schedule
    for m in self._read_schedule(ts)]
  File "/home/charles/pypath/huey/api.py", line 164, in inner
    wrap_exception(exc_class)
  File "/home/charles/pypath/huey/utils.py", line 18, in wrap_exception
    raise new_exc_class('%s: %s' % (exc_class.__name__, exc))
ScheduleReadException: ConnectionError: Error 111 connecting to localhost:6379. Connection refused.

[2016-07-12 22:14:17,067] ERROR:huey.consumer.Worker:Worker-1:Error reading from queue
Traceback (most recent call last):
  File "/home/charles/pypath/huey/consumer.py", line 94, in loop
    task = self.huey.dequeue()
  File "/home/charles/pypath/huey/api.py", line 225, in dequeue
    message = self._dequeue()
  File "/home/charles/pypath/huey/api.py", line 164, in inner
    wrap_exception(exc_class)
  File "/home/charles/pypath/huey/utils.py", line 18, in wrap_exception
    raise new_exc_class('%s: %s' % (exc_class.__name__, exc))
QueueReadException: ConnectionError: Error 111 connecting to localhost:6379. Connection refused.
[2016-07-12 22:14:17,068] DEBUG:huey.consumer.Worker:Worker-1:No messages, sleeping for: 1.63665373929
[2016-07-12 22:14:17,070] ERROR:huey.consumer.Worker:Worker-2:Error reading from queue
Traceback (most recent call last):
  File "/home/charles/pypath/huey/consumer.py", line 94, in loop
    task = self.huey.dequeue()
  File "/home/charles/pypath/huey/api.py", line 225, in dequeue
    message = self._dequeue()
  File "/home/charles/pypath/huey/api.py", line 164, in inner
    wrap_exception(exc_class)
  File "/home/charles/pypath/huey/utils.py", line 18, in wrap_exception
    raise new_exc_class('%s: %s' % (exc_class.__name__, exc))
QueueReadException: ConnectionError: Error 111 connecting to localhost:6379. Connection refused.
[2016-07-12 22:14:17,070] DEBUG:huey.consumer.Worker:Worker-2:No messages, sleeping for: 1.63665373929
[2016-07-12 22:14:18,709] ERROR:huey.consumer.Worker:Worker-1:Error reading from queue
Traceback (most recent call last):
  File "/home/charles/pypath/huey/consumer.py", line 94, in loop
    task = self.huey.dequeue()
  File "/home/charles/pypath/huey/api.py", line 225, in dequeue
    message = self._dequeue()
  File "/home/charles/pypath/huey/api.py", line 164, in inner
    wrap_exception(exc_class)
  File "/home/charles/pypath/huey/utils.py", line 18, in wrap_exception
    raise new_exc_class('%s: %s' % (exc_class.__name__, exc))
QueueReadException: ConnectionError: Error 111 connecting to localhost:6379. Connection refused.
[2016-07-12 22:14:18,710] DEBUG:huey.consumer.Worker:Worker-1:No messages, sleeping for: 1.88215180019
[2016-07-12 22:14:18,710] ERROR:huey.consumer.Worker:Worker-2:Error reading from queue
Traceback (most recent call last):
  File "/home/charles/pypath/huey/consumer.py", line 94, in loop
    task = self.huey.dequeue()
  File "/home/charles/pypath/huey/api.py", line 225, in dequeue
    message = self._dequeue()
  File "/home/charles/pypath/huey/api.py", line 164, in inner
    wrap_exception(exc_class)
  File "/home/charles/pypath/huey/utils.py", line 18, in wrap_exception
    raise new_exc_class('%s: %s' % (exc_class.__name__, exc))
QueueReadException: ConnectionError: Error 111 connecting to localhost:6379. Connection refused.
[2016-07-12 22:14:18,711] DEBUG:huey.consumer.Worker:Worker-2:No messages, sleeping for: 1.88215180019
[2016-07-12 22:14:20,596] ERROR:huey.consumer.Worker:Worker-1:Error reading from queue
Traceback (most recent call last):
  File "/home/charles/pypath/huey/consumer.py", line 94, in loop
    task = self.huey.dequeue()
  File "/home/charles/pypath/huey/api.py", line 225, in dequeue
    message = self._dequeue()
  File "/home/charles/pypath/huey/api.py", line 164, in inner
    wrap_exception(exc_class)
  File "/home/charles/pypath/huey/utils.py", line 18, in wrap_exception
    raise new_exc_class('%s: %s' % (exc_class.__name__, exc))
QueueReadException: ConnectionError: Error 111 connecting to localhost:6379. Connection refused.
[2016-07-12 22:14:20,597] DEBUG:huey.consumer.Worker:Worker-1:No messages, sleeping for: 2.16447457022
[2016-07-12 22:14:20,599] ERROR:huey.consumer.Worker:Worker-2:Error reading from queue
Traceback (most recent call last):
  File "/home/charles/pypath/huey/consumer.py", line 94, in loop
    task = self.huey.dequeue()
  File "/home/charles/pypath/huey/api.py", line 225, in dequeue
    message = self._dequeue()
  File "/home/charles/pypath/huey/api.py", line 164, in inner
    wrap_exception(exc_class)
  File "/home/charles/pypath/huey/utils.py", line 18, in wrap_exception
    raise new_exc_class('%s: %s' % (exc_class.__name__, exc))
QueueReadException: ConnectionError: Error 111 connecting to localhost:6379. Connection refused.
[2016-07-12 22:14:20,599] DEBUG:huey.consumer.Worker:Worker-2:No messages, sleeping for: 2.16447457022
[2016-07-12 22:14:22,765] ERROR:huey.consumer.Worker:Worker-1:Error reading from queue
Traceback (most recent call last):
  File "/home/charles/pypath/huey/consumer.py", line 94, in loop
    task = self.huey.dequeue()
  File "/home/charles/pypath/huey/api.py", line 225, in dequeue
    message = self._dequeue()
  File "/home/charles/pypath/huey/api.py", line 164, in inner
    wrap_exception(exc_class)
  File "/home/charles/pypath/huey/utils.py", line 18, in wrap_exception
    raise new_exc_class('%s: %s' % (exc_class.__name__, exc))
QueueReadException: ConnectionError: Error 111 connecting to localhost:6379. Connection refused.
[2016-07-12 22:14:22,766] DEBUG:huey.consumer.Worker:Worker-1:No messages, sleeping for: 2.48914575575
[2016-07-12 22:14:22,768] ERROR:huey.consumer.Worker:Worker-2:Error reading from queue
Traceback (most recent call last):
  File "/home/charles/pypath/huey/consumer.py", line 94, in loop
    task = self.huey.dequeue()
  File "/home/charles/pypath/huey/api.py", line 225, in dequeue
    message = self._dequeue()
  File "/home/charles/pypath/huey/api.py", line 164, in inner
    wrap_exception(exc_class)
  File "/home/charles/pypath/huey/utils.py", line 18, in wrap_exception
    raise new_exc_class('%s: %s' % (exc_class.__name__, exc))
QueueReadException: ConnectionError: Error 111 connecting to localhost:6379. Connection refused.
[2016-07-12 22:14:22,768] DEBUG:huey.consumer.Worker:Worker-2:No messages, sleeping for: 2.48914575575
[2016-07-12 22:14:25,258] ERROR:huey.consumer.Worker:Worker-1:Error reading from queue
Traceback (most recent call last):
  File "/home/charles/pypath/huey/consumer.py", line 94, in loop
    task = self.huey.dequeue()
  File "/home/charles/pypath/huey/api.py", line 225, in dequeue
    message = self._dequeue()
  File "/home/charles/pypath/huey/api.py", line 164, in inner
    wrap_exception(exc_class)
  File "/home/charles/pypath/huey/utils.py", line 18, in wrap_exception
    raise new_exc_class('%s: %s' % (exc_class.__name__, exc))
QueueReadException: ConnectionError: Error 111 connecting to localhost:6379. Connection refused.
[2016-07-12 22:14:25,258] DEBUG:huey.consumer.Worker:Worker-1:No messages, sleeping for: 2.86251761911
[2016-07-12 22:14:25,262] ERROR:huey.consumer.Worker:Worker-2:Error reading from queue
Traceback (most recent call last):
  File "/home/charles/pypath/huey/consumer.py", line 94, in loop
    task = self.huey.dequeue()
  File "/home/charles/pypath/huey/api.py", line 225, in dequeue
    message = self._dequeue()
  File "/home/charles/pypath/huey/api.py", line 164, in inner
    wrap_exception(exc_class)
  File "/home/charles/pypath/huey/utils.py", line 18, in wrap_exception
    raise new_exc_class('%s: %s' % (exc_class.__name__, exc))
QueueReadException: ConnectionError: Error 111 connecting to localhost:6379. Connection refused.
[2016-07-12 22:14:25,262] DEBUG:huey.consumer.Worker:Worker-2:No messages, sleeping for: 2.86251761911
[2016-07-12 22:14:28,122] ERROR:huey.consumer.Worker:Worker-1:Error reading from queue
Traceback (most recent call last):
  File "/home/charles/pypath/huey/consumer.py", line 94, in loop
    task = self.huey.dequeue()
  File "/home/charles/pypath/huey/api.py", line 225, in dequeue
    message = self._dequeue()
  File "/home/charles/pypath/huey/api.py", line 164, in inner
    wrap_exception(exc_class)
  File "/home/charles/pypath/huey/utils.py", line 18, in wrap_exception
    raise new_exc_class('%s: %s' % (exc_class.__name__, exc))
QueueReadException: ConnectionError: Error 111 connecting to localhost:6379. Connection refused.
[2016-07-12 22:14:28,122] DEBUG:huey.consumer.Worker:Worker-1:No messages, sleeping for: 3.29189526198
[2016-07-12 22:14:28,129] ERROR:huey.consumer.Worker:Worker-2:Error reading from queue
Traceback (most recent call last):
  File "/home/charles/pypath/huey/consumer.py", line 94, in loop
    task = self.huey.dequeue()
  File "/home/charles/pypath/huey/api.py", line 225, in dequeue
    message = self._dequeue()
  File "/home/charles/pypath/huey/api.py", line 164, in inner
    wrap_exception(exc_class)
  File "/home/charles/pypath/huey/utils.py", line 18, in wrap_exception
    raise new_exc_class('%s: %s' % (exc_class.__name__, exc))
QueueReadException: ConnectionError: Error 111 connecting to localhost:6379. Connection refused.
[2016-07-12 22:14:28,129] DEBUG:huey.consumer.Worker:Worker-2:No messages, sleeping for: 3.29189526198
[2016-07-12 22:14:31,419] DEBUG:huey.consumer.Worker:Worker-1:No messages, sleeping for: 3.78567955128
[2016-07-12 22:14:31,424] DEBUG:huey.consumer.Worker:Worker-2:No messages, sleeping for: 3.78567955128
[2016-07-12 22:14:35,208] DEBUG:huey.consumer.Worker:Worker-1:No messages, sleeping for: 4.35353148397
[2016-07-12 22:14:35,213] DEBUG:huey.consumer.Worker:Worker-2:No messages, sleeping for: 4.35353148397
^C[2016-07-12 22:14:36,928] INFO:huey.consumer:MainThread:Shutting down
[2016-07-12 22:14:37,029] INFO:huey.consumer:MainThread:Consumer exiting.
```

**Comment by coleifer** [comment]
The workers in this case are not blocking but polling, which may account for the difference?

**Comment by martinmo** [comment]
I just realized that the reported behavior is very closely related to my other reported issue #164. Using the latest Huey installed from Git (i.e., including the fix for #164), everything now works as expected. Everything meaning: Django's run_huey command as well as the "simple" example. Thanks for your quick response!
